src/extensions/compression/deflate.rs

```diff
@@ -0,0 +1,289 @@
+const SERVER_NO_CONTEXT_TAKEOVER: &str = "server_no_context_takeover";
+const CLIENT_MAX_WINDOW_BITS: &str = "client_max_window_bits";
+const SERVER_MAX_WINDOW_BITS: &str = "server_max_window_bits";
+
+const TRAILER: [u8; 4] = [0x00, 0x00, 0xff, 0xff];
+
+/// Error from `permessage-deflate` extension.
+#[derive(Debug, Error)]
+pub enum DeflateError {
+    /// Compress failed
+    #[error("failed to compress: {0}")]
+    Compress(std::io::Error),
+    /// Decompress failed
+    #[error("failed to decompress: {0}")]
+    Decompress(std::io::Error),
+}
+
+// Parameters `server_max_window_bits` and `client_max_window_bits` are not supported for now
+// because custom window size requires `flate2/zlib` feature.
+// TODO Configs for how the server accepts these offers.
+/// Configurations for `permessage-deflate` Per-Message Compression Extension.
+#[derive(Clone, Copy, Debug)]
+pub struct DeflateConfig {
+    /// Compression level.
+    pub compression: Compression,
+    /// Request the peer server not to use context takeover.
+    pub server_no_context_takeover: bool,
+    /// Hint that context takeover is not used.
+    pub client_no_context_takeover: bool,
+}
+
+impl Default for DeflateConfig {
+    fn default() -> Self {
+        Self {
+            compression: Compression::default(),
+            server_no_context_takeover: false,
+            client_no_context_takeover: false,
+        }
+    }
+}
+
+impl DeflateConfig {
+    pub(crate) fn name(&self) -> &str {
+        PER_MESSAGE_DEFLATE
+    }
+
+    /// Value for `Sec-WebSocket-Extensions` request header.
+    pub(crate) fn negotiation_offers(&self) -> HeaderValue {
+        let mut offers = Vec::new();
+        if self.server_no_context_takeover {
+            offers.push(Param::new(SERVER_NO_CONTEXT_TAKEOVER));
+        }
+        if self.client_no_context_takeover {
+            offers.push(Param::new(CLIENT_NO_CONTEXT_TAKEOVER));
+        }
+        to_header_value(&offers)
+    }
+
+    // This can be used for `WebSocket::from_raw_socket_with_compression`.
+    /// Returns negotiation response based on offers and `DeflateContext` to manage per message compression.
+    pub fn negotiation_response(&self, extensions: &str) -> Option<(HeaderValue, DeflateContext)> {
+        // Accept the first valid offer for `permessage-deflate`.
+        // A server MUST decline an extension negotiation offer for this
+        // extension if any of the following conditions are met:
+        // * The negotiation offer contains an extension parameter not defined
+        //   for use in an offer.
+        // * The negotiation offer contains an extension parameter with an
+        //   invalid value.
+        // * The negotiation offer contains multiple extension parameters with
+        //   the same name.
+        // * The server doesn't support the offered configuration.
+        'outer: for (_, offer) in
+            extensions::parse_header(extensions).iter().filter(|(k, _)| k == self.name())
```

Comment:
Ok, I think I got the meaning of the function, but I believe that it could be simplified (and so also there would be no need for the `'outer`) if the function accepts the input for the deflate only, e.g. it would be something like:

```
fn generate_response(input_params: &[Param]) -> Response {
    let mut (param1, param2, ...) = (false, false);
    for param in params {
    }
    return Response::from(param1, param2, ...)
}
```

Does it make any sense?